compiler/rustc_codegen_ssa/src/back/link.rs

```diff
@@ -391,6 +400,44 @@ fn link_rlib<'a, B: ArchiveBuilder<'a>>(
     ab
 }
 
+/// Extract all symbols defined in raw-dylib libraries, collated by library name.
+///
+/// If we have multiple extern blocks that specify symbols defined in the same raw-dylib library,
+/// then the CodegenResults value contains one NativeLib instance for each block.  However, the
+/// linker appears to expect only a single import library for each library used, so we need to
+/// collate the symbols together by library name before generating the import libraries.
+fn collate_raw_dylibs(used_libraries: &[NativeLib]) -> Vec<(Symbol, Vec<DllImport>)> {
+    let mut dylib_table: FxHashMap<Symbol, FxHashSet<Symbol>> = FxHashMap::default();
+
+    for lib in used_libraries {
+        if lib.kind == NativeLibKind::RawDylib {
+            let name = lib.name.unwrap_or_else(||
+                bug!("`link` attribute with kind = \"raw-dylib\" and no name should have caused error earlier")
+            );
+            dylib_table
+                .entry(name)
+                .or_default()
+                .extend(lib.dll_imports.iter().map(|import| import.name));
+        }
+    }
+
+    // FIXME: when we add support for ordinals, fix this to propagate ordinals.  Also figure out
+    // what we should do if we have two DllImport values with the same name but different
+    // ordinals.
+    dylib_table
+        .into_iter()
+        .map(|(lib_name, imported_names)| {
+            (
+                lib_name,
+                imported_names
+                    .iter()
+                    .map(|name| DllImport { name: *name, ordinal: None })
+                    .collect::<Vec<_>>(),
+            )
+        })
+        .collect::<Vec<_>>()
```

Comment:
Nit: the hash table here introduces non-deterministic compilation (imports can be written in different random order).
Perhaps it's better to sort the resulting vector in a predictable way before returning it.